### ISSUE TITLE
LogReuseWaitStatus enum is incomplete

### ISSUE THREAD

**Issue by petervandivier** [microsoft/sqlmanagementobjects#105]
`LogReuseWaitStatus` lists 10 valid statuses. This matches [the documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.smo.logreusewaitstatus) however both of these are incomplete and should include all possible values. Note the below example (from [dbatools' `Get-DbaDatabase`](https://github.com/dataplat/dbatools/blob/development/functions/Get-DbaDatabase.ps1)) where this enum is surfaced but shows inconsistent values due to the incomplete enum.

![image](https://user-images.githubusercontent.com/10898767/181023879-067e7f5f-e081-4010-b9a6-251abaf1ddbe.png)
![image](https://user-images.githubusercontent.com/10898767/181023900-3f6c1997-88f7-4ac6-ac23-8a596e455979.png) 

https://github.com/microsoft/sqlmanagementobjects/blob/98093604e9ede97aa2a9b066279f0bd9d9557c4f/src/Microsoft/SqlServer/Management/SqlEnum/enumstructs.cs#L2194-L2216

Using the [`sys.databases` documentation](https://docs.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-databases-transact-sql) as a reference, types 10 through 17 are missing; although how the "Internal Use Only" types are exposed is perhaps not for me to suggest

| log_reuse_wait | log_reuse_wait_desc | 
| --- | --- | 
| 10 | For internal use only | 
| 11 | For internal use only | 
| 12 | For internal use only | 
| 13 | Oldest page | 
| 14 | Other | 
| 16 | XTP_CHECKPOINT | 
| 17 | sLog scanning |

**Comment by shueybubbles** [comment]
I think this is another example of `enum` abuse, like #102 lists for various strings.
SMO shouldn't use an enum when SQL can change the set of valid values even between CUs for the same major version.  Maybe for v17 we should change the properties that report this enum type to `int` instead and change `enum LogReuseWaitStatus` to `static public class LogReuseWaitStatus` and the current enum values to constants.

**Comment by ben-thul** [comment]
@shueybubbles - just so I'm following, would I still be able to do string comparisons against whatever is returned for the Database.LogReuseWaitStatus property?

**Comment by shueybubbles** [comment]
for this case I'd keep it as a number not a string.

**Comment by ben-thul** [comment]
Can you walk me through the particulars of consuming/using that number? The beauty of the current state (i.e. having an enum) is that the value that comes back is immediately consumable by a human. If I have to start hard-coding magic numbers into code that needs to use this value (e.g. `if (db.LogReuseWaitStatus == 6) { /* 6 = "replication" */`) that feels like a regression from a user experience standpoint.

**Comment by shueybubbles** [comment]
``` C#
public enum foo { x = 1 }

```
=>
``` C#
public static class foo 
{
    public const int x = 1;
}
```

**Comment by ben-thul** [comment]
I promise I'm not trying to be difficult; I just don't understand. In your envisioned end state, what does this code change to:

```powershell
if ($db.LogReuseWaitStatus -eq 'Replication') {
   # do something
}
```

That is, right now, I'm not calling anything explicitly on the enum returned as the property `LogReuseWaitStatus` of the `$db` object. So I think I need something more than "`foo.x` is a valid symbol".
---
Additionally, how does moving this to a static class address the problem of being able to sync with the ground truth of what is in the SQL Server product? Using this PR as an example, you'd presumably have to apprise *something* of the meaning the value coming back from SQL (e.g. "13" means "Oldest page").

**Comment by shueybubbles** [comment]
in powershell I guess it might be

```PS

if ($db.LogReuseWaitStatus -eq Microsoft.SqlServer.Management.Smo.LogReuseWaitStatus.Replication) {
  # do something
}

```

The difference between using an int and enum is that your application could understand what the new value means before SMO does and you could use the value without being blocked.
If SMO eventually adds a constant for the new value, it's a convenience but not required. When it's an `enum` you are essentially forced to wait for SMO to add the new value before you can reference it.

In C# code nobody would have been using the string representation of the enum value so there'd be no code change needed.

**Comment by shueybubbles** [comment]
I don't think we can use the strings from `log_reuse_wait_desc` because they get localized.

**Comment by shueybubbles** [comment]
Does powershell support using the fully qualified symbol of the enum value instead of the name of the enum value for the comparison? IE could you use `$db.LogReuseStatus -eq LogReuseStatus.Replication` in your code today?

**Comment by ben-thul** [comment]
> Does powershell support using the fully qualified symbol of the enum value instead of the name of the enum value for the comparison?

Kind of. It looks something like:

```powershell
$db.LogReuseWaitStatus -eq [Microsoft.SqlServer.Management.Smo.LogReuseWaitStatus]::Replication
```

In later versions of powershell, you can get around having to fully qualify the enum name with `using namespace Microsoft.SqlServer.Management.Smo;` (or some prefix).

---

And I think it's coming together for me. What you're proposing is to compare the property of the database object (e.g. `$db.LogReuseWaitStatus`) against a class property (e.g. `[LogReuseStatus].Replication`). That feels okay to me. Thanks for the explanation!
